### ISSUE TITLE
BeamSearchDecoder cell state never changed 

### ISSUE THREAD

**Issue by chenghuige** [tensorflow/tensorflow#9904]
tf version '1.1.0-rc2'
It looks BeamSearchDecoder never used next_cell_state, but always using the inital cell state.
In beam_search_decoder.py 423
        next_cell_state = nest.map_structure(
            self._maybe_split_batch_beams,
            next_cell_state, self._cell.state_size)
But next_cell_state is never used later, since just pass   state to _beam_search_step function where state.cell_state is never changed 
    beam_search_output, beam_search_state = _beam_search_step(
          time=time,
          logits=cell_outputs,
          beam_state=state,
          batch_size=batch_size,
          beam_width=beam_width,
          end_token=end_token,
          length_penalty_weight=length_penalty_weight)

**Comment by shuuki4** [comment]
I wrote the code using BeamSearchDecoder, and struggling to solve the issue that the first step's output looks quite OK but later steps' output seems strange (similar to first step's output). This seems to be the reason why..

**Comment by chenghuige** [comment]
Also we need to reorder new cell state according to next_beam_ids.

**Comment by chenghuige** [comment]
Also there are still bugs if fixing above problem. For BeamSearchDecoder design, it is dynamic and stop when all facing <EOS>, but is it good for doing beam search, I think done beams should not competing with non finished beams.  Take a small example, compare the result of out graph beam search(beam_width 2) and beam search by BeamSearchDecoder:
for out graph beam search I got:
0 [936, 2] 古代/<EOS> 0.000512714519986 -7.57579 -7.57579 [-3.9758704, -3.599921]
1 [1, 2] <UNK>/<EOS> 0.000276204156072 -8.19437 -8.19437 [-4.6221266, -3.5722442]
for BeamSearchDecoder, I print predicted_ids, parent_ids and sequence_lengths,  and I transpose final beam search output predicted_ids [0, 2, 1] to print the top 2 best paths.
predicted_ids:  
[[[936   1]
  [  8   2]
  [  2   2]]]
parent_ids: 
[[[0 0]
  [0 0]
  [1 0]]]
sequence_lengths:
[[3 2]]
[936   2   2] 古代/<EOS> 0.000512714 -7.57579140233
[936   2  -1] 古代/<EOS> 0.000134506 -8.91390228447

Which is not as good as out graph result, you see the path [1, 2] will be ignored since we have 
[936, 8] [936, 2] occupy all 2 beam_width positions...
And the final result has [936, 2, 1] [936, 2, -1] is also not correct.

**Comment by concretevitamin** [comment]
@ebrevdo mind taking a look?

**Comment by wirehack** [comment]
I seem to encounter the same problem. Some help would be much appreciated.

**Comment by ebrevdo** [comment]
Looking.

**Comment by ebrevdo** [comment]
Thanks for finding this rather egregious bug.  We've got a bugfix ready and
hope to submit it today, and merge it into the TF1.2 branch.  We'll also be
adding some better (functional) unit tests in the coming week.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assigneee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 354 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 369 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 384 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 399 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 414 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 429 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 444 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by chenghuige** [comment]
closing this issue, seems eager mode can do better beam search.

**Comment by ebrevdo** [comment]
To be clear, we fixed this bug last year
